### PR TITLE
Yashveer/feture bagging eda

### PR DESCRIPTION
**Pull request by yashveersinghsohi**
# Basic Info
This branch aims to explore the benefits of grouping features together in a meaningful fashion so that we can build ensemble like models around feature functions. 

For eg: we can build models that predict qualities like `politeness`, `emotion`, `question` etc. and then use the predictions for these aggregated features to build the final predictions of efficiency and team performance. 

That allows us to make very powerful and interpretable predictions about team performance and also it allows us to make very specific statements about what qualities impact performance the most.

# Files Changed
The code is written in the `feature_engine/testing/ipython_notebooks/feature_bagging_eda.ipynb` and `feature_engine/testing/ipython_notebooks/feature_bagging_nn.ipynb` notebooks.

In addition to this, the trained models from these notebooks are stored in the following directory: `feature_engine/testing/ipython_notebooks/feature_bagging_nn_models`.

# Code Description
There are 2 broad ways in which we have aimed to group features together:
- **Grouping features that have high correlations with each other.**
  - This mechanism defeats the purpose of interpretability as the sub models built using features collected in a group do not measure/predict a meaningful qualitative quality about team communication.
- **Grouping features that have similar function or features that measure a similar quantity.**
  - This is exactly what we want as now we can use sub models to predict qualities like `politeness`, `emotion` etc, and then use these measures of `politeness`, `emotion` etc to predict team performance.
  - These grouping are available in a more description format [here](https://docs.google.com/spreadsheets/d/15FoHAjtScQbOZvclH3GUKYI4YyUAS431_FGzBvr9J1s/edit?usp=sharing), and are stored in a minimalistic fashion in the repo here - `feature_engine/testing/ipython_notebooks/feature_groupings_06_20_23.xlsx`.

The challenge across both these approaches is that we do not have targets for any of our sub-models. i.e. we do not have a measure of `politeness` to use as ground truth for the politeness sub-model. We essentially have a set of inputs that predict a set of latent (unobserved) qualities of a conversation, which then in turn predict the performance of the team. 

This is very similar to a 1 layer artificial neural network structure, where we have a set of inputs feeding into a set of hidden neurons which then are used to predict the final target. In addition to this, in a neural network we do not know what the ground truths of the hidden layer should be - which is the same situation that we face in the grouping exercise.

The notebook `feature_engine/testing/ipython_notebooks/feature_bagging_nn_models` builds a bunch of neural network style architectures that aim to solve the problem of missing intermediate targets. 
- **Model 1: Linear Regression Expressed as a Neural Network**
  - This is essentially Linear Regression expressed as a Neural Network to get an understanding of how to define and use Pytorch models, datasets and training paradigms.
- **Model 2: Fully Connected Neural Network**
  - This is a standard traditional neural network where all input features are connected to all neurons from the hidden layer. Here each neuron represents a feature group and hence we have 15 neurons in the hidden layer.
- **Model 3: Sparsely Connected Neural Network**
  - Since the goal is to have a group of features (for eg: the group that measures `politeness`) used to predict just one quality of conversation (`politeness` in this case) and then use the measures of these *qualities* to predict the target, we can not use a fully connected network. We need to remove some connections - features that are being connected to neurons from a different group (for eg: the connection between a `politeness` feature and the  `emotion` neuron would not be correct and should be removed).
  - There is no straightforward way to do this in the standard deep learning packages. We have implemented 2 workarounds:
  - In this workaround, we allow fully connected networks where the first layer has the same number of neurons as the number of features , but we mask out the neurons that do not make sense according to the feature groups to result in 15 neurons.
  - The disadvantage here is that, in the remaining 15 neurons, we still get a fully connected network. Essentially we need some way to mask out weights instead of neurons.
- **Model 4: Sparsely Connected Neural Network V2**
  - In this model we use the same architecture as model 2, but before training in each epoch, we zero out the weights that are not needed, and we make sure that the gradient descent updates are made in a way that the weights zero-ed out are not affected by the updates.
  - Hence, in this approach, we essentially have removed the connections that were not needed as they are zero throughout the training and validation stages of the model training lifecycle. 

Despite this extensive exercise, the performance we get out of these novel architectures is not comparable to the one we get from just throwing all the features to a Random Forests or an XGBoost model. Maybe that is due to the limited availability of rows in our dataset and due to the fact that we have not done extensive hyperparameter tuning yet while defining the model architecture (learning rate, early stopping, etc).

Hence the recommendation is to start integrating the 4th model into the model pipeline and try and improve its performance in the model optimization workstream.

### CODE REVIEW

**Review by xehu — APPROVED** [review]
Per discussion -- this PR is self-contained, so let's push it in and continue with exploring the NN.
